File: ldk-bukkit-particle/src/main/java/com/lgou2w/ldk/bukkit/particle/Particle.java
/*
 * Copyright (C) 2016-2021 The lgou2w <lgou2w@hotmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.lgou2w.ldk.bukkit.particle;

import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Objects;

public final class Particle {
  @NotNull private final String namespacedKey;
  @NotNull private final String namespace;
  @NotNull private final String key;
  // legacy Minecraft 1.12 and before
  @Nullable private final Integer id;

  @Contract("null, _, _ -> fail; _, null, _ -> fail")
  Particle(String namespace, String key, @Nullable Integer id) {
    this.namespace = Objects.requireNonNull(namespace, "namespace");
    this.key = Objects.requireNonNull(key, "key");
    this.namespacedKey = namespace + ':' + key;
    this.id = id;
  }

  @Contract("null, _ -> fail")
  Particle(String key, @Nullable Integer id) {
    this("minecraft", key, id);
  }

  @Contract("null -> fail")
  Particle(String key) {
    this("minecraft", key, null);
  }

  @NotNull
  public String getNamespace() {
    return namespace;
  }

  @NotNull
  public String getKey() {
    return key;
  }

  @NotNull
  public String getNamespacedKey() {
    return namespacedKey;
  }

  @Deprecated
  public int getId() {
    return id != null ? id : -1;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Particle particle = (Particle) o;
    return namespace.equals(particle.namespace) && key.equals(particle.key);
  }

  @Override
  public int hashCode() {
    return Objects.hash(namespace, key);
  }

  @Override
  public String toString() {
    return "Particle{" + namespacedKey + '}';
  }

  public final static Particle POOF = new Particle("poof", 0);
  public final static Particle EXPLOSION = new Particle("explosion", 1);
  public final static Particle EXPLOSION_EMITTER = new Particle("explosion_emitter", 2);
  public final static Particle FIREWORK = new Particle("firework", 3);
  public final static Particle BUBBLE = new Particle("bubble", 4);
  public final static Particle SPLASH = new Particle("splash", 5);
  public final static Particle FISHING = new Particle("fishing", 6);
  public final static Particle UNDERWATER = new Particle("underwater", 7);
  @Deprecated // Removed in Minecraft 1.13. Replaced -> UNDERWATER
  public final static Particle UNDERWATER_DEPTH = new Particle("underwater", 8);
  public final static Particle CRIT = new Particle("crit", 9);
  public final static Particle ENCHANTED_HIT = new Particle("enchanted_hit", 10);
  public final static Particle SMOKE = new Particle("smoke", 11);
  public final static Particle LARGE_SMOKE = new Particle("large_smoke", 12);
  public final static Particle EFFECT = new Particle("effect", 13);
  public final static Particle INSTANT_EFFECT = new Particle("instant_effect", 14);
  public final static Particle ENTITY_EFFECT = new Particle("entity_effect", 15);
  public final static Particle AMBIENT_ENTITY_EFFECT = new Particle("ambient_entity_effect", 16);
  public final static Particle WITCH = new Particle("witch", 17);
  public final static Particle DRIPPING_WATER = new Particle("dripping_water", 18);
  public final static Particle DRIPPING_LAVA = new Particle("dripping_lava", 19);
  public final static Particle ANGRY_VILLAGER = new Particle("angry_villager", 20);
  public final static Particle HAPPY_VILLAGER = new Particle("happy_villager", 21);
  public final static Particle MYCELIUM = new Particle("mycelium", 22);
  public final static Particle NOTE = new Particle("note", 23);
  public final static Particle PROTAL = new Particle("protal", 24);
  public final static Particle ENCHANT = new Particle("enchant", 25);
  public final static Particle FLAME = new Particle("flame", 26);
  public final static Particle LAVA = new Particle("lava", 27);
  @Deprecated // Removed in Minecraft 1.13. Replaced -> BARRIER
  public final static Particle FOOTSTEP = new Particle("barrier", 28);
  public final static Particle CLOUD = new Particle("cloud", 29);
  public final static Particle DUST = new Particle("dust", 30);
  public final static Particle ITEM_SNOWBALL = new Particle("item_snowball", 31);
  @Deprecated // Removed in Minecraft 1.13. Replaced -> ITEM_SNOWBALL
  public final static Particle ITEM_SNOW_SHOVEL = new Particle("item_snowball", 32);
  public final static Particle ITEM_SLIME = new Particle("item_slime", 33);
  public final static Particle HEART = new Particle("heart", 34);
  public final static Particle BARRIER = new Particle("barrier", 35);
  public final static Particle ITEM = new Particle("item", 36);
  public final static Particle BLOCK = new Particle("block", 37);
  @Deprecated // Removed in Minecraft 1.13. Replaced -> BLOCK
  public final static Particle BLOCK_DUST = new Particle("block", 38);
  public final static Particle RAIN = new Particle("rain", 39);
  @Deprecated // Removed in Minecraft 1.13. Replaced -> ITEM
  public final static Particle ITEM_TAKE = new Particle("item", 40);
  public final static Particle ELDER_GUARDIAN = new Particle("elder_guardian", 41);

  // Minecraft 1.9 or later
  public final static Particle DRAGON_BREATH = new Particle("dragon_breath", 42);
  public final static Particle END_ROD = new Particle("end_rod", 43);
  public final static Particle DAMAGE_INDICATOR = new Particle("damage_indicator", 44);
  public final static Particle SWEEP_ATTACK = new Particle("sweep_attack", 45);

  // Minecraft 1.10 or later
  public final static Particle FALLING_DUST = new Particle("falling_dust", 46);

  // Minecraft 1.11 or later
  public final static Particle TOTEM_OF_UNDYING = new Particle("totem_of_undying", 47);
  public final static Particle SPIT = new Particle("spit", 48);

  // Minecraft 1.13 or later. No longer support ID!
  public final static Particle SQUID_INK = new Particle("squid_ink");
  public final static Particle BUBBLE_POP = new Particle("bubble_pop");
  public final static Particle CURRENT_DOWN = new Particle("current_down");
  public final static Particle BUBBLE_COLUMN_UP = new Particle("bubble_column_up");
  public final static Particle NAUTILUS = new Particle("nautilus");
  public final static Particle DOLPHIN = new Particle("dolphin");

  // Minecraft 1.14 or later
  public final static Particle SNEEZE = new Particle("sneeze");
  public final static Particle CAMPFIRE_COSY_SMOKE = new Particle("campfire_cosy_smoke");
  public final static Particle CAMPFIRE_SIGNAL_SMOKE = new Particle("campfire_signal_smoke");
  public final static Particle COMPOSTER = new Particle("composter");
  public final static Particle FLASH = new Particle("flash");
  public final static Particle FALLING_LAVA = new Particle("falling_lava");
  public final static Particle LANDING_LAVA = new Particle("landing_lava");
  public final static Particle FALLING_WATER = new Particle("falling_water");

  // Minecraft 1.15 or later
  public final static Particle DRIPPING_HONEY = new Particle("dripping_honey");
  public final static Particle FALLING_HONEY = new Particle("falling_honey");
  public final static Particle LANDING_HONEY = new Particle("landing_honey");
  public final static Particle FALLING_NECTAR = new Particle("falling_nectar");

  // Minecraft 1.16 or later
  public final static Particle SOUL_FIRE_FLAME = new Particle("soul_fire_flame");
  public final static Particle ASH = new Particle("ash");
  public final static Particle CRIMSON_SPORE = new Particle("crimson_spore");
  public final static Particle WARPED_SPORE = new Particle("warped_spore");
  public final static Particle SOUL = new Particle("soul");
  public final static Particle DRIPPING_OBSIDIAN_TEAR = new Particle("dripping_obsidian_tear");
  public final static Particle FALLING_OBSIDIAN_TEAR = new Particle("falling_obsidian_tear");
  public final static Particle LANDING_OBSIDIAN_TEAR = new Particle("landing_obsidian_tear");
  public final static Particle REVERSE_PORTAL = new Particle("reverse_portal");
  public final static Particle WHITE_ASH = new Particle("white_ash");

  // Instance methods

  public void display(@NotNull Location center, double range, float offsetX, float offsetY, float offsetZ, float speed, int count, @Nullable Object data) {
    ParticleFactory.sendParticle(this, center, range, offsetX, offsetY, offsetZ, speed, count, data);
  }

  public void display(@NotNull Location center, double range, float offsetX, float offsetY, float offsetZ, float speed, @Nullable Object data) {
    display(center, range, offsetX, offsetY, offsetZ, speed, 1, data);
  }

  public void display(@NotNull Location center, double range, float offsetX, float offsetY, float offsetZ, float speed, int count) {
    display(center, range, offsetX, offsetY, offsetZ, speed, count, null);
  }

  public void display(@NotNull Location center, double range, float offsetX, float offsetY, float offsetZ, float speed) {
    display(center, range, offsetX, offsetY, offsetZ, speed, 1, null);
  }

  public void display(@NotNull Location center, double range, float offsetX, float offsetY, float offsetZ, @Nullable Object data) {
    display(center, range, offsetX, offsetY, offsetZ, 0f, 1, data);
  }

  public void display(@NotNull Location center, double range, float offsetX, float offsetY, float offsetZ) {
    display(center, range, offsetX, offsetY, offsetZ, 0f, 1, null);
  }

  public void display(@NotNull Location center, double range, @Nullable Object data) {
    display(center, range, 0f, 0f, 0f, 0f, 1, data);
  }

  public void display(@NotNull Location center, double range) {
    display(center, range, 0f, 0f, 0f, 0f, 1, null);
  }

  public void display(@NotNull Location center, @Nullable Object data) {
    display(center, 32d, 0f, 0f, 0f, 0f, 1, data);
  }

  public void display(@NotNull Location center) {
    display(center, 32d, 0f, 0f, 0f, 0f, 1, null);
  }

  public void display(@Nullable Player sender, @NotNull Location center, float offsetX, float offsetY, float offsetZ, float speed, int count, @Nullable Object data) {
    ParticleFactory.sendParticle(this, sender, center, offsetX, offsetY, offsetZ, speed, count, data);
  }

  public void display(@Nullable Player sender, @NotNull Location center, float offsetX, float offsetY, float offsetZ, float speed, @Nullable Object data) {
    display(sender, center, offsetX, offsetY, offsetZ, speed, 1, data);
  }

  public void display(@Nullable Player sender, @NotNull Location center, float offsetX, float offsetY, float offsetZ, float speed, int count) {
    display(sender, center, offsetX, offsetY, offsetZ, speed, count, null);
  }

  public void display(@Nullable Player sender, @NotNull Location center, float offsetX, float offsetY, float offsetZ, float speed) {
    display(sender, center, offsetX, offsetY, offsetZ, speed, 1, null);
  }

  public void display(@Nullable Player sender, @NotNull Location center, float offsetX, float offsetY, float offsetZ, @Nullable Object data) {
    display(sender, center, offsetX, offsetY, offsetZ, 0f, 1, data);
  }

  public void display(@Nullable Player sender, @NotNull Location center, float offsetX, float offsetY, float offsetZ) {
    display(sender, center, offsetX, offsetY, offsetZ, 0f, 1, null);
  }

  public void display(@Nullable Player sender, @NotNull Location center, @Nullable Object data) {
    display(sender, center, 0f, 0f, 0f, 0f, 1, data);
  }

  public void display(@Nullable Player sender, @NotNull Location center) {
    display(sender, center, 0f, 0f, 0f, 0f, 1, null);
  }

  public void display(@NotNull List<Player> receivers, @NotNull Location center, float offsetX, float offsetY, float offsetZ, float speed, int count, @Nullable Object data) {
    ParticleFactory.sendParticle(this, receivers, center, offsetX, offsetY, offsetZ, speed, count, data);
  }

  public void display(@NotNull List<Player> receivers, @NotNull Location center, float offsetX, float offsetY, float offsetZ, float speed, @Nullable Object data) {
    display(receivers, center, offsetX, offsetY, offsetZ, speed, 1, data);
  }

  public void display(@NotNull List<Player> receivers, @NotNull Location center, float offsetX, float offsetY, float offsetZ, float speed, int count) {
    display(receivers, center, offsetX, offsetY, offsetZ, speed, count, null);
  }

  public void display(@NotNull List<Player> receivers, @NotNull Location center, float offsetX, float offsetY, float offsetZ, float speed) {
    display(receivers, center, offsetX, offsetY, offsetZ, speed, 1, null);
  }

  public void display(@NotNull List<Player> receivers, @NotNull Location center, float offsetX, float offsetY, float offsetZ, @Nullable Object data) {
    display(receivers, center, offsetX, offsetY, offsetZ, 0f, 1, data);
  }

  public void display(@NotNull List<Player> receivers, @NotNull Location center, float offsetX, float offsetY, float offsetZ) {
    display(receivers, center, offsetX, offsetY, offsetZ, 0f, 1, null);
  }

  public void display(@NotNull List<Player> receivers, @NotNull Location center, @Nullable Object data) {
    display(receivers, center, 0f, 0f, 0f, 0f, 1, data);
  }

  public void display(@NotNull List<Player> receivers, @NotNull Location center) {
    display(receivers, center, 0f, 0f, 0f, 0f, 1, null);
  }
}

File: ldk-bukkit/ldk-bukkit-common/src/main/kotlin/com/lgou2w/ldk/bukkit/particle/Particle.kt
/*
 * Copyright (C) 2016-2019 The lgou2w <lgou2w@hotmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.lgou2w.ldk.bukkit.particle

import com.lgou2w.ldk.bukkit.version.API
import com.lgou2w.ldk.bukkit.version.Draft
import com.lgou2w.ldk.bukkit.version.Level
import com.lgou2w.ldk.bukkit.version.MinecraftBukkitVersion
import com.lgou2w.ldk.common.Valuable
import org.bukkit.Location
import org.bukkit.entity.Player
import java.util.Collections
import java.util.HashMap
import kotlin.math.abs

/**
 * ## Particle (粒子效果)
 *
 * @see [Valuable]
 * @see [ParticleFactory]
 * @author lgou2w
 */
enum class Particle(
        private val internal: Int,
        /**
         * * The old version name of this particle.
         * * 此粒子的旧版本名称.
         */
        val legacy: String,
        /**
         * * The new version name of this particle.
         * * 此粒子的新版本名称.
         */
        val type: String
) : Valuable<Int> {

    /**
     * * 如果粒子由 `@Deprecated` 注解那么代表在 `Minecraft 1.13` 中被移除.
     *      不过依然是能够用的, 仅仅只是兼容层. 只是不会记住 Id 而已.
     *
     * * 由 `@API` 注解的粒子说明是对应版本才加入的, 使用时注意.
     *
     * 目前 `Minecraft 1.12.2` 为止, 可用的粒子效果一共为 49 种. [SPIT] 结束.
     * 在 `Minecraft 1.13` 中分别移除了 [UNDERWATER_DEPTH]、[FOOTSTEP]、[ITEM_SNOW_SHOVEL]、[BLOCKDUST]、[TAKE]，
     *      不过没关系的, 这些在 1.13 依然能够使用, LDK 已经为这些做好兼容.
     */

    POOF(34, "explosion_normal", "poof"),
    EXPLOSION(19, "explosion_large", "explosion"),
    EXPLOSION_EMITTER(18, "explosion_huge", "explosion_emitter"),
    FIREWORK(21, "fireworks_spark", "firework"),
    BUBBLE(4, "water_bubble", "bubble"),
    SPLASH(43, "water_splash", "splash"),
    FISHING(22, "water_wake", "fishing"),
    UNDERWATER(42, "suspended", "underwater"),
    @Deprecated("Removed in Minecraft 1.13, Automatically adapt if used.", replaceWith = ReplaceWith("UNDERWATER"))
    UNDERWATER_DEPTH("suspended_depth", -42),
    CRIT(6, "crit", "crit"),
    ENCHANTED_HIT(14, "crit_magic", "enchanted_hit"),
    SMOKE(37, "smoke_normal", "smoke"),
    LARGE_SMOKE(30, "smoke_large", "large_smoke"),
    EFFECT(12, "spell", "effect"),
    INSTANT_EFFECT(26, "spell_instant", "instant_effect"),
    ENTITY_EFFECT(17, "spell_mob", "entity_effect"),
    AMBIENT_ENTITY_EFFECT(0, "spell_mob_ambient", "ambient_entity_effect"),
    WITCH(44, "spell_witch", "witch"),
    DRIPPING_WATER(10, "drip_water", "dripping_water"),
    DRIPPING_LAVA(9, "drip_lava", "dripping_lava"),
    ANGRY_VILLAGER(1, "villager_angry", "angry_villager"),
    HAPPY_VILLAGER(24, "villager_happy", "happy_villager"),
    MYCELIUM(32, "town_aura", "mycelium"),
    NOTE(33, "note", "note"),
    PORTAL(35, "portal", "portal"),
    ENCHANT(15, "enchantment_table", "enchant"),
    FLAME(23, "flame", "flame"),
    LAVA(31, "lava", "lava"),
    @Deprecated("Removed in Minecraft 1.13, Automatically adapt if used.")
    FOOTSTEP("footstep", -2),
    CLOUD(5, "cloud", "cloud"),
    DUST(11, "red_dust", "dust"),
    ITEM_SNOWBALL(29, "snowball", "item_snowball"),
    @Deprecated("Removed in Minecraft 1.13, Automatically adapt if used.", replaceWith = ReplaceWith("ITEM_SNOWBALL"))
    ITEM_SNOW_SHOVEL("snow_shovel", -29),
    ITEM_SLIME(28, "slime", "item_slime"),
    HEART(25, "heart", "heart"),
    BARRIER(2, "barrier", "barrier"),
    ITEM(27, "item_crack", "item"),
    BLOCK(3, "block_crack", "block"),
    @Deprecated("Removed in Minecraft 1.13, Automatically adapt if used.", replaceWith = ReplaceWith("BLOCK"))
    BLOCKDUST("blockdust", -3),
    RAIN(36, "water_drop", "rain"),
    @Deprecated("Removed in Minecraft 1.13, Automatically adapt if used.", replaceWith = ReplaceWith("ITEM"))
    TAKE("item_take", -27),
    ELDER_GUARDIAN(13, "mob_appearance", "elder_guardian"),

    @API(Level.Minecraft_V1_9) DRAGON_BREATH(8, "dragon_breath", "dragon_breath"),
    @API(Level.Minecraft_V1_9) END_ROD(16, "end_rod", "end_rod"),
    @API(Level.Minecraft_V1_9) DAMAGE_INDICATOR(7, "damage_indicator", "damage_indicator"),
    @API(Level.Minecraft_V1_9) SWEEP_ATTACK(40, "sweep_attack", "sweep_attack"),
    @API(Level.Minecraft_V1_10) FALLING_DUST(20, "falling_dust", "falling_dust"),
    @API(Level.Minecraft_V1_11) TOTEM_OF_UNDYING(41, "totem", "totem_of_undying"),
    @API(Level.Minecraft_V1_11) SPIT(38, "spit", "spit"),

    @API(Level.Minecraft_V1_13) SQUID_INK(39, "squid_ink", "squid_ink"),
    @API(Level.Minecraft_V1_13) BUBBLE_POP(45, "bubble_pop", "bubble_pop"),
    @API(Level.Minecraft_V1_13) CURRENT_DOWN(46, "current_down", "current_down"),
    @API(Level.Minecraft_V1_13) BUBBLE_COLUMN_UP(47, "bubble_column_up", "bubble_column_up"),
    @API(Level.Minecraft_V1_13) NAUTILUS(48, "nautilus", "nautilus"),
    @API(Level.Minecraft_V1_13) DOLPHIN(49, "dolphin", "dolphin"),

    @API(Level.Minecraft_V1_14) SNEEZE(50, "sneeze", "sneeze"),
    @API(Level.Minecraft_V1_14) CAMPFIRE_COSY_SMOKE(51, "campfire_cosy_smoke", "campfire_cosy_smoke"),
    @API(Level.Minecraft_V1_14) CAMPFIRE_SIGNAL_SMOKE(52, "campfire_signal_smoke", "campfire_signal_smoke"),
    @API(Level.Minecraft_V1_14) COMPOSTER(53, "composter", "composter"),
    @API(Level.Minecraft_V1_14) FLASH(54, "flash", "flash"),
    @API(Level.Minecraft_V1_14) FALLING_LAVA(55, "falling_lava", "falling_lava"),
    @API(Level.Minecraft_V1_14) LANDING_LAVA(56, "landing_lava", "landing_lava"),
    @API(Level.Minecraft_V1_14) FALLING_WATER(57, "falling_water", "falling_water"),

    @Draft @Deprecated("Minecraft 1.15 Draft") DRIPPING_HONEY(58, "dripping_honey", "dripping_honey"),
    @Draft @Deprecated("Minecraft 1.15 Draft") FALLING_HONEY(59, "falling_honey", "falling_honey"),
    @Draft @Deprecated("Minecraft 1.15 Draft") LANDING_HONEY(60, "landing_honey", "landing_honey"),
    @Draft @Deprecated("Minecraft 1.15 Draft") FALLING_NECTAR(61, "falling_nectar", "falling_nectar"),
    ;

    constructor(
            legacy: String,
            equivalent: Int
    ) : this(equivalent, legacy, legacy)

    override val value : Int
        get() = if (MinecraftBukkitVersion.isV113OrLater)
            @Suppress("DEPRECATION")
            (abs(internal))
        else
            ordinal

    companion object {

        @JvmStatic var COMPATIBILITIES : Array<Particle> private set
        @JvmStatic private val ID_MAP : Map<Int, Particle>
        @JvmStatic private val NAME_MAP : Map<String, Particle>

        // internal < 0 => skip
        // Particle > SPIT && CURRENT < 1.13 => skip

        init {
            val isV113OrLater = MinecraftBukkitVersion.isV113OrLater
            val idMap = HashMap<Int, Particle>()
            val nameMap = HashMap<String, Particle>()
            values().forEach { particle ->
                idMap[particle.value] = particle
                nameMap[particle.legacy] = particle
                nameMap[particle.type] = particle
            }
            ID_MAP = Collections.unmodifiableMap(idMap)
            NAME_MAP = Collections.unmodifiableMap(nameMap)
            COMPATIBILITIES = ID_MAP.values
                .filter { (isV113OrLater && it.internal >= 0) || (!isV113OrLater && it.ordinal <= SPIT.ordinal) }
                .toTypedArray()
        }

        @JvmStatic
        fun fromId(id: Int): Particle
                = ID_MAP[id] ?: throw IllegalArgumentException("Unsupported particle effect type: $id")

        @JvmStatic
        fun fromName(name: String): Particle
                = NAME_MAP[name] ?: throw IllegalArgumentException("Unsupported particle effect type: $name")
    }

    /**************************************************************************
     *
     * Particle Send Extended
     *
     **************************************************************************/

    /**
     * @since LDK 0.1.8-rc
     */
    @JvmOverloads
    fun display(
            center: Location,
            range: Double = 32.0,
            offsetX: Float = 0f,
            offsetY: Float = 0f,
            offsetZ: Float = 0f,
            speed: Float = 0.0f,
            count: Int = 1,
            data: Any? = null
    ) = ParticleFactory.sendParticle(this, center, range, offsetX, offsetY, offsetZ, speed, count, data)

    /**
     * @since LDK 0.1.8-rc
     */
    @JvmOverloads
    fun display(
            sender: Player?,
            center: Location,
            offsetX: Float = 0f,
            offsetY: Float = 0f,
            offsetZ: Float = 0f,
            speed: Float = 0.0f,
            count: Int = 1,
            data: Any? = null
    ) = ParticleFactory.sendParticle(sender, this, center, offsetX, offsetY, offsetZ, speed, count, data)

    /**
     * @since LDK 0.1.8-rc
     */
    @JvmOverloads
    fun display(
            players: List<Player>,
            center: Location,
            offsetX: Float = 0f,
            offsetY: Float = 0f,
            offsetZ: Float = 0f,
            speed: Float = 0.0f,
            count: Int = 1,
            data: Any? = null
    ) = ParticleFactory.sendParticle(players, this, center, offsetX, offsetY, offsetZ, speed, count, data)
}
